subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/AttributeMatchingVariantSelector.java

```diff
@@ -101,130 +96,78 @@ public ResolvedArtifactSet select(ResolvedVariantSet producer, Factory factory)
-            VariantDefinition definition = result.getRight();
-            return factory.asTransformed(variant, definition, dependenciesResolver, transformedVariantFactory);
+
+        if (transformedVariants.size() == 1) {
+            TransformedVariant result = transformedVariants.get(0);
+            return factory.asTransformed(result.getRoot(), result.getVariantChain(), dependenciesResolver, transformedVariantFactory);
         }
 
-        if (!candidates.isEmpty()) {
-            throw new AmbiguousTransformException(producer.asDescribable().getDisplayName(), componentRequested, candidates);
+        if (!transformedVariants.isEmpty()) {
+            throw new AmbiguousTransformException(producer.asDescribable().getDisplayName(), componentRequested, transformedVariants);
         }
 
         if (ignoreWhenNoMatches) {
             return ResolvedArtifactSet.EMPTY;
         }
+
         throw new NoMatchingVariantSelectionException(producer.asDescribable().getDisplayName(), componentRequested, variants, matcher, DescriberSelector.selectDescriber(componentRequested, schema));
     }
 
-    private List<Pair<ResolvedVariant, MutableConsumerVariantMatchResult.ConsumerVariant>> tryDisambiguate(AttributeMatcher matcher, List<Pair<ResolvedVariant, MutableConsumerVariantMatchResult.ConsumerVariant>> candidates, ImmutableAttributes componentRequested, AttributeMatchingExplanationBuilder explanationBuilder) {
-        candidates = disambiguateWithSchema(matcher, candidates, componentRequested, explanationBuilder);
-
+    private List<TransformedVariant> tryDisambiguate(AttributeMatcher matcher, List<TransformedVariant> candidates, ImmutableAttributes componentRequested, AttributeMatchingExplanationBuilder explanationBuilder) {
+        List<TransformedVariant> matches = matcher.matches(candidates, componentRequested, explanationBuilder);
         if (candidates.size() == 1) {
-            return candidates;
+            return matches;
         }
-
-        if (candidates.size() == 2) {
-            // Short circuit logic when only 2 candidates
-            return compareCandidates(matcher, candidates.get(0), candidates.get(1))
-                .map(Collections::singletonList)
-                .orElse(candidates);
+        if (matches.size() > 0 && matches.size() < candidates.size()) {
+            candidates = matches;
         }
 
-        List<Pair<ResolvedVariant, MutableConsumerVariantMatchResult.ConsumerVariant>> shortestTransforms = Lists.newArrayListWithExpectedSize(candidates.size());
-        candidates.sort(Comparator.comparingInt(candidate -> candidate.right.depth));
+        List<TransformedVariant> differentTransforms = new ArrayList<>(1);
 
-        // Need to remember if a further element was matched by an earlier one, no need to consider it then
-        boolean[] hasBetterMatch = new boolean[candidates.size()];
+        TransformedVariant last = candidates.get(candidates.size() - 1);
```

Comment:
Is there anything meaningful about starting with the last candidate?

Reply:
Nope. This is strictly a simplification of the existing algorithm with all logic for comparing depth removed. The old implementation simplifies down to always adding the last element and then comparing the rest of the items to the last. 

I've left this behavior for compatibility. Though, using the last element is strictly arbitrary.